sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileSourceStrategy.scala

```diff
@@ -79,8 +79,16 @@ object FileSourceStrategy extends Strategy with Logging {
         ExpressionSet(normalizedFilters.filter(_.references.subsetOf(partitionSet)))
       logInfo(s"Pruning directories with: ${partitionKeyFilters.mkString(",")}")
 
+      val prunedFsRelation = fsRelation.location match {
```

Comment:
That's correct. Let's wait for #14750 then. My feeling here is that schema inference over a subset of files is not acceptable since then the schema can vary with the partition predicate.

For the broadcast join optimization, I think that is a larger issue that should not affect the design of this pr.